pkg/steps/multi_stage/multi_stage.go

```diff
@@ -216,19 +225,28 @@ func (s *multiStageTestStep) run(ctx context.Context) error {
 	}
 	observerContext, cancel := context.WithCancel(ctx)
 	observerDone := make(chan struct{})
+	defer close(observerDone)
 	go s.runObservers(observerContext, ctx, observers, observerDone)
```

Comment:
If I understand correctly the source of the flakes you explained in the description of this PR, it looks like it's because we passing the same context in this go routine. Instead, we can pass a new context with a timeout, instead of implementing the whole timeout idea with a `time.Ticker` in the `tryWaitObserversOrTimeout` method.